### DISCUSSION TITLE
TCP support on 15.4 (Linux platform)

### DISCUSSION THREAD

**Discussion by hastigondaliya** [openthread/openthread#8442] [Q&A] [answered]
Hello,

I'm trying to make TCP connection between imx8m host(with FreeRTOS and Linux platform) and nrf52840. imx8m host is connected with Thread supported module over SPI.

TCP connection successfully established with FreeRTOS platform but with Linux Platform host send TCP reset after established the connection.
 
<img width="960" alt="image" src="https://user-images.githubusercontent.com/118960681/203759329-53f78ead-7edb-4550-8ce2-f88cc4403fe5.png">

<img width="960" alt="image" src="https://user-images.githubusercontent.com/118960681/203754960-14662772-0cb0-4c56-85df-269b3f4c61fe.png">

Observation is that "**OPENTHREAD_CONFIG_PLATFORM_NETIF_ENABLE**" is required for Linux platform and enabled only for daemon usecase. This interface is disabled for Firecrest FreeRTOS platform. 

Also, the TCP reset message is being propagated via NETIF socket, thats causing TCP connection reset.

**Attachment:** OTA and syslog
[tcp_reset_logs.zip](https://github.com/openthread/openthread/files/10083482/tcp_reset_logs.zip)

Any suggestion/hint would be very helpful.

**Comment by jwhui** [comment, marked as answer]
Check to make sure that [`OPENTHREAD_CONFIG_TCP_ENABLE`](https://github.com/openthread/openthread/blob/27b43146a931f2d53380f9e895223ce18e5a83d3/src/core/config/ip6.h#L172) is set to 0 when compiling for Linux host. Otherwise OpenThread's TCP implementation may conflict with Linux's TCP implementation.

OTBR disable's OpenThread's TCP implementation: https://github.com/openthread/ot-br-posix/blob/6fe884b9df5a8918665ac260c3a75e93c44ae602/third_party/openthread/CMakeLists.txt#L103

**Comment by hastigondaliya** [comment]
Hello @jwhui,

Thank you for your valuable inputs.

I need to enable the TCP support for ot-daemon POSIX platform. I'm using https://github.com/openthread/openthread openthread repo. If OPENTHREAD_CONFIG_TCP_ENABLE set to 0 in openthread from https://github.com/openthread/openthread/blob/27b43146a931f2d53380f9e895223ce18e5a83d3/src/core/config/ip6.h#L172 then ot-ctl won't allow to execute any TCP command. 

Could you please suggest what can be done in this case?

**Reply by jwhui** [reply]
In general, when using POSIX platform, you want to use the TCP stack provided by the POSIX platform. You won't be able to use OT CLI to drive TCP in this case.

**Comment by hastigondaliya** [comment]
Hello @jwhui,

Thank you for your input.

We wanted to enable OT TCPLP stack on the POSIX platform, and if we are running ot in daemon, we are seeing TCP reset sent out by POSIX TCP stack.

We observed that whenever OT TCPLP enabled for POSIX platform, it will have two TCP stacks - openthread and POSIX TCP stack. So, any TCP packet received from Reference device, this packet is processed by both stacks (openthread tcplp and POSIX tcp stacks). To avoid this dual processing, we would like to propose the below solution.

We can restrict the TCP messages to pass from openthread stack as done for ICMP6 and UDP messages in **ProcessReceiveCallback()**, whenever TCPLP feature is enabled.

Please find attached patch for this solution.
[enable_ot_tcp_with_posix_platform.zip](https://github.com/openthread/openthread/files/10174853/enable_ot_tcp_with_posix_platform.zip)

Can this solution be integrated to opensource openthread stack?

**Reply by jwhui** [reply]
Using the host's TCP stack allows applications on the host to use standard networking services (e.g. socket API) to communicate over Thread. Using OpenThread's TCP stack would not allow applications on the host to use the standard POSIX APIs. If this is the use case you intend to support, then your proposed patch should be fine. However, I'm hesitant to accept this into OpenThread main branch since we generally want to encourage integration into the host's networking stack.

**Reply by hastigondaliya** [reply]
Hi @jwhui,

"Using OpenThread's TCP stack would not allow applications on the host to use the standard POSIX APIs. If this is the use case you intend to support, then your proposed patch should be fine."

Yes, This would be the expected use case and the intent of the feature flag OPENTHREAD_CONFIG_TCP_ENABLE. Because if POSIX platform intends to use host/POSIX's TCP stack then it is expected that openthread TCP stack will not be used. That's the reason in our patch file we have kept the changes under feature flag OPENTHREAD_CONFIG_TCP_ENABLE. So that this changes are applicable only when openthread TCP stack is used and not when POSIX TCP stack used.

So basically this patch will not discourage using host networking stack but this is to fix TCP connection failure issue that has pop up due to dual processing of the same packet by both POSIX and openthread TCP stacks. This patch will make sure TCP connection works fine either of the stacks (POSIX/openthread TCP stacks), with/without feature flag OPENTHREAD_CONFIG_TCP_ENABLE enabled

**Reply by jwhui** [reply]
Just curious - can you provide a use case / motivation for using OpenThread's TCPLP stack instead of the POSIX host?

**Reply by hastigondaliya** [reply]
Hi @jwhui,

This is just for convenience of internal testing that we are using this config as some of our test scripts are using ot-cli interface.

**Reply by jwhui** [reply]
OK, that makes sense. Would you like to submit a PR with the proposed change?

**Reply by hastigondaliya** [reply]
Hello @jwhui,

Yes, we would like to submit a PR.
